nix/cabal-project.nix

```diff
@@ -0,0 +1,159 @@
+          plutus-metatheory.components.tests.test1.preCheck =
+            let
+              cmp = config.hsPkgs.plutus-metatheory.components.tests.test1;
+              deps = cmp.executableToolDepends;
+            in
+            ''PATH=${lib.makeBinPath deps }:$PATH'';
+
+          # FIXME: Somehow this is broken even with setting the path up as above
+          plutus-metatheory.components.tests.test2.doCheck = false;
+
+          # plutus-metatheory needs agda with the stdlib around for the custom setup
+          # I can't figure out a way to apply this as a blanket change for all the
+          # components in the package, oh well
+          plutus-metatheory.components.library.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.exes.plc-agda.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.tests.test1.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.tests.test2.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+          plutus-metatheory.components.tests.test3.build-tools = [ repoRoot.nix.agda-with-stdlib ];
+
+          plutus-core.components.benchmarks.update-cost-model = {
+            build-tools = [ repoRoot.nix.r-with-packages ];
+          };
+
+          plutus-core.components.benchmarks.cost-model-test = {
+            build-tools = [ repoRoot.nix.r-with-packages ];
+          };
+        };
+      }
+      (lib.mkIf (config.compiler-nix-name != "ghc8107") {
+        packages = {
+          # -Werror for CI
+          # Only enable on the newer compilers. We don't care about warnings on the old ones,
+          # and sometimes it's hard to be warning free on all compilers, e.g. the unused
+          # packages warning is bad in 8.10.7
+          # (https://gitlab.haskellib.org/ghc/ghc/-/merge_requests/6130)
+
+          # Werror everything.
+          # This is a pain, see https://github.com/input-output-hk/haskell.nix/issues/519
+          plutus-benchmark.ghcOptions = [ "-Werror" ];
+          plutus-conformance.ghcOptions = [ "-Werror" ];
+          plutus-core.ghcOptions = [ "-Werror" ];
+          plutus-ledger-api.ghcOptions = [ "-Werror" ];
+          # FIXME: has warnings in generated code
+          #plutus-metatheory.package.ghcOptions = "-Werror";
+          plutus-tx.ghcOptions = [ "-Werror" ];
+          plutus-tx-plugin.ghcOptions = [ "-Werror" ];
+          # This package's tests require doctest, which generates Haskell source
+          # code. However, it does not add derivation strategies in said code,
+          # which will fail the build with -Werror. Furthermore, barring an
+          # upstream fix, there's nothing we can do about it other than
+          # disabling -Werror on it.
+          prettyprinter-configurable.ghcOptions = [ "-Werror" ];
+        };
+      })
+    ];
+  };
+
+
+  project = lib.iogx.mkHaskellProject {
```

Comment:
What you get is the `iogx` overlay added to each `haskell-nix.cabalProject` variant.
In addition you get a `devShell` for each project variant, which includes the `pre-commit-check`.
Then you get the `combined-haddock` derivation.
plutus and plutus-apps use the combined-haddock to build the RTD site, so the RTD site needs to reference the haskell project somehow, since it also needs to add some scripts to the shell.

Separating the RTD into its own `lib.iogx.mkReadTheDocsSite` means more work for end user.

Reply:
> plutus and plutus-apps use the combined-haddock to build the RTD site, so the RTD site needs to reference the haskell project somehow, since it also needs to add some scripts to the shell.
> Separating the RTD into its own lib.iogx.mkReadTheDocsSite means more work for end user.

IDK, I guess I just don't like having a big "do everything" function, but I guess that's kind of the point. How do I fix anything in the RTD site if it's not working properly?

Reply:
e.g. it looks like you've just taken a frozen snapshot of the python packages we happen to currently be using for Sphinx, but we might want to add more - there's 0 guarantee that that's the final set. So what would I do then?